File: code/tc/common/src/vmtemplate.cpp
/*******************************************************************************
* Copyright (c) 2011，中兴通讯股份有限公司，All rights reserved.
* 
* 文件名称：VmTemplate.cpp
* 文件标识：见配置管理计划书
* 内容摘要：VmTemplate类的实现文件
* 当前版本：1.0 
* 作    者：Bob
* 完成日期：2011年7月28日
*
* 修改记录1：
*    修改日期：2011/7/28
*    版 本 号：V1.0
*    修 改 人：Bob
*    修改内容：创建
*******************************************************************************/

#include "vmtemplate.h"
#include "sky.h"

namespace zte_tecs{

/******************************************************************************/
void VmTemplate::Modify(const ApiVtModifyData &modify_msg )
{
    _name      = modify_msg.vt_name;
    _is_public = modify_msg.is_public;
    
    _description = modify_msg.cfg_info.base_info.description;
    _vcpu    = modify_msg.cfg_info.base_info.vcpu;
    _tcu     = modify_msg.cfg_info.base_info.tcu;
    _memory  = modify_msg.cfg_info.base_info.memory;
        
    // image, disk
    _images.clear();
    VmBaseImage   vm_base_image(INVALID_OID,
                                modify_msg.cfg_info.base_info.machine.target,
                                modify_msg.cfg_info.base_info.machine.position,
                                modify_msg.cfg_info.base_info.machine.id,
                                modify_msg.cfg_info.base_info.machine.reserved_backup,
                                modify_msg.cfg_info.base_info.machine.type);
    _images.push_back(vm_base_image);

    _disks.clear();
    vector<ApiVmDisk>::const_iterator vec_it_disk; 
    for (vec_it_disk  = modify_msg.cfg_info.base_info.disks.begin();
         vec_it_disk != modify_msg.cfg_info.base_info.disks.end();
         vec_it_disk++)
    {
        if ((INVALID_OID == vec_it_disk->id)
            &&(DISK_TYPE_DISK == vec_it_disk->type))
        {
            VmBaseDisk vm_base_disk(INVALID_OID,
                                        vec_it_disk->target,
                                        vec_it_disk->position,
                                        vec_it_disk->bus,
                                        vec_it_disk->size,
                                        vec_it_disk->fstype);
                                        
            _disks.push_back(vm_base_disk);
        }
        else if((INVALID_OID != vec_it_disk->id)
               &&(DISK_TYPE_DISK != vec_it_disk->type))
        {
            VmBaseImage   vm_base_image(INVALID_OID,
                                    vec_it_disk->target,
                                    vec_it_disk->position,
                                    vec_it_disk->id,
                                    vec_it_disk->reserved_backup,
                                    vec_it_disk->type);
                                    
            _images.push_back(vm_base_image);
        } 
    }


    // context
    _context._str_map.clear();
    map<string, string>::const_iterator  it_context;
    for (it_context  = modify_msg.cfg_info.base_info.context.begin();
         it_context != modify_msg.cfg_info.base_info.context.end();
         it_context++)
    {           
        _context._id = INVALID_OID; 
        _context._str_map.insert(make_pair(it_context->first, it_context->second));
    }
        
    // 调度条件
    _qualifications._qua.clear();
    vector<ApiVmExpression>::const_iterator vec_it_exp; 
    for (vec_it_exp = modify_msg.cfg_info.base_info.qualifications.begin();
             vec_it_exp !=  modify_msg.cfg_info.base_info.qualifications.end();
             vec_it_exp++)
    {
        Expression  exp(vec_it_exp->key,
                            vec_it_exp->op,
                            vec_it_exp->value);
            
        _qualifications._qua.push_back(exp);
    }       
    
    vector<VmBaseNic>      tmp_nic;   
    vector<ApiVmNic>::const_iterator  it_nic_new;
      
    for(it_nic_new = modify_msg.cfg_info.base_info.nics.begin();
        it_nic_new != modify_msg.cfg_info.base_info.nics.end();
        it_nic_new++)
    {
        // 1. nic
        VmBaseNic   vm_base_nic(INVALID_OID,
                            it_nic_new->nic_index,
                            it_nic_new->sriov,
                            it_nic_new->loop,
                            it_nic_new->logic_network_id,
                            it_nic_new->model,
                            "");
 
        tmp_nic.push_back(vm_base_nic);
    }  
    
    _nics.clear();   
    _nics = tmp_nic;     
       
   _virt_type           = modify_msg.cfg_info.base_info.virt_type;
   _root_device         = modify_msg.cfg_info.base_info.root_device;
   _kernel_command      = modify_msg.cfg_info.base_info.kernel_command;
   _bootloader          = modify_msg.cfg_info.base_info.bootloader;
   _raw_data            = modify_msg.cfg_info.base_info.rawdata;
   _appointed_cluster   = modify_msg.cfg_info.base_info.cluster;
   _appointed_host      = modify_msg.cfg_info.base_info.host;
       
   _watchdog            = modify_msg.cfg_info.base_info.vm_wdtime;
   _vnc_passwd          = modify_msg.cfg_info.base_info.vnc_passwd;
   _enable_coredump = modify_msg.cfg_info.base_info.enable_coredump;
   _enable_serial       = modify_msg.cfg_info.base_info.enable_serial;
   _hypervisor          = modify_msg.cfg_info.base_info.hypervisor;
   _enable_livemigrate  = modify_msg.cfg_info.base_info.enable_livemigrate;

    _devices.assign(modify_msg.cfg_info.base_info.devices.begin(), modify_msg.cfg_info.base_info.devices.end());
}


/******************************************************************************/

}
	

File: tools/new_ci/base_function_module/interface/base_interface/rpc/vmtemplate.py
#!/usr/bin/python
# -*- coding: gb2312 -*-
#*******************************************************************************
# Copyright (c) 2012，中兴通讯股份有限公司，All rights reserved.
# 
# 文件名称：vmcfg.py
# 测试内容：vmcfg模块xmlrpc接口封装
# 当前版本：1.0 
# 作    者：李孝成
# 完成日期：2012/03/24
#
# 修改记录1：
#    修改日期：2012/03/24
#    版 本 号：V1.0
#    修 改 人：李孝成
#    修改内容：创建
#*******************************************************************************/
from rpc import tecs_rpc

#****************************************************************************#
#                             常量定义                                       # 
#****************************************************************************#
VT_START_INDEX = 0L
VT_QUERY_COUNT = 50L
VT_DISABLE = bool(0)
VT_ENABLE = bool(1)
QUERY_USER_CREAT_VT = 1
QUERY_USER_VISIBLE_VT = 2
QUERY_BY_VT_NAME = 3
QUERY_FIRST_NIC = 0
QUERY_SECOND_NIC = 1
NO_VT_NAME = ""
#****************************************************************************#
class vmtemplate:
    ''' tecs vmtemplate methods '''
    def __init__(self, host, session):
        self.api = tecs_rpc(host)
        self.apisvr = self.api.server
        self.session = session
        
            
    def allocate(self,cfg):
        result = self.apisvr.tecs.vmtemplate.allocate(self.session,cfg) 
        if result[0] != 0:
            print "failed to create vmt! error code = %d" % (result[0])
            print "error string = %s" % result[1]
            return False
        else:
            print "success to create vmt %s!" % cfg["vt_name"]
            return True

    def set(self,cfg):
        result = self.apisvr.tecs.vmtemplate.set(self.session,cfg) 
        if result[0] != 0:
            print "failed to set vmt! error code = %d" % (result[0])
            print "error string = %s" % result[1]
            return False
        else:
            print "success to set vmt %s!" % cfg["vt_name"]
            return True           
            
    def del_vmt(self,vt_name):
        result = self.apisvr.tecs.vmtemplate.delete(self.session,vt_name) 
        if result[0] != 0:
            print "failed to del vmt error code = %d" % (result[0])
            print "error string = %s" % result[1]
            return False
        else:
            print "success to del vmt !"
            return True
           
    def publish_vmt(self,vt_name):
        result = self.apisvr.tecs.vmtemplate.publish(self.session,vt_name,bool(1)) 
        if result[0] != 0:
            print "failed to publish vmt error code = %d" % (result[0])
            print "error string = %s" % result[1]
            return False
        else:
            print "success to publish vmt !"
            return True
            
    def unpublish_vmt(self,vt_name):
        result = self.apisvr.tecs.vmtemplate.publish(self.session,vt_name,bool(0)) 
        if result[0] != 0:
            print "failed to unpublish vmt error code = %d" % (result[0])
            print "error string = %s" % result[1]
            return False
        else:
            print "success to unpublish vmt !"
            return True
            
    def show_vmt_info(self,type,vt_name):
        '''显示指定的虚拟机配置信息  '''
        result = self.apisvr.tecs.vmtemplate.query(self.session,VT_START_INDEX,VT_QUERY_COUNT,type,vt_name) 
        if result[0] != 0:
            print "failed to query vmt information, type %d! error code =%d" % (type,result[0])
            print "error string = %s" % result[1]
            return False
        else:
            if len(result[3]) == 0:
                print "no vmt type %d!" % type
                return True;
            
            vtcfg_list = result[3]
            for v in vtcfg_list:
                print "==========================="
                print "vmt id = %d" % v["vt_id"]
                print "vmt name = %s" % v["vt_name"]
                print "public is = %d" % v["is_public"]
                print "uid = %s" % v["uid"]
                print "owner = %s" % v["user_name"]
                print "created at = %s" % v["create_time"]
                
                #print "vm name = %s" % v["cfg_info"]["vm_name"]
                #print "project_name = %s" % v["cfg_info"]["project_name"]
              
                
                print "description = %s" % v["cfg_info"]["base_info"]["description"]
                print "vcpu = %d" % v["cfg_info"]["base_info"]["vcpu"]
                print "tcu = %d" % v["cfg_info"]["base_info"]["tcu"]
                print "memory = %d" % v["cfg_info"]["base_info"]["memory"]
                print "root_device = %s" % v["cfg_info"]["base_info"]["root_device"]
                print "kernel_command = %s" % v["cfg_info"]["base_info"]["kernel_command"]
                print "bootloader = %s" % v["cfg_info"]["base_info"]["bootloader"]
                print "virt_type = %s" % v["cfg_info"]["base_info"]["virt_type"]
                print "hypervisor = %s" % v["cfg_info"]["base_info"]["hypervisor"]
                print "rawdata = %s" % v["cfg_info"]["base_info"]["rawdata"]
                print "assigned cluster = %s" % v["cfg_info"]["base_info"]["cluster"]
                print "assigned host = %s" % v["cfg_info"]["base_info"]["host"]
                print "vm_wdtime = %d" % v["cfg_info"]["base_info"]["vm_wdtime"]
                print "vnc_passwd = %s" % v["cfg_info"]["base_info"]["vnc_passwd"]
                print "enable_coredump = %d" % v["cfg_info"]["base_info"]["enable_coredump"]
                print "enable_serial = %d" % v["cfg_info"]["base_info"]["enable_serial"]
                
                print "---machine info:---------------" 
                print "machine virtual disk %d " % v["cfg_info"]["base_info"]["machine"]["id"]
                print "position: %d " % v["cfg_info"]["base_info"]["machine"]["position"]
                print "disk size: %d " % v["cfg_info"]["base_info"]["machine"]["size"]
                print "bus: %s " % v["cfg_info"]["base_info"]["machine"]["bus"]
                print "disk type: %s" % v["cfg_info"]["base_info"]["machine"]["type"]
                print "target: %s" % v["cfg_info"]["base_info"]["machine"]["target"]
                print "fstype: %s" % v["cfg_info"]["base_info"]["machine"]["fstype"]
                print "----machine info---------------" 
              
                disk_list = v["cfg_info"]["base_info"]["disks"]
                for d in disk_list:
                    print "---virtual disk %d ------------" % d["id"]
                    print "position: %d " % d["position"]
                    print "disk size: %d " % d["size"]
                    print "bus: %s " % d["bus"]
                    print "disk type: %s" % d["type"]
                    print "target: %s" % d["target"]
                    print "fstype: %s" % d["fstype"]
                    print "-------------------------------" 

                nic_list = v["cfg_info"]["base_info"]["nics"]
                for n in nic_list:
                    print "----virtual network card %d ---" % n["nic_index"]
                    print "sriov %d " % n["sriov"]
                    print "loop %d " % n["loop"]
                    print "logic_network_id: %s " % n["logic_network_id"]
                    print "model: %s " % n["model"]
                    print "-------------------------------"

                q_list = v["cfg_info"]["base_info"]["qualifications"]
                for q in q_list:
                    print "-------------------------------"
                    print "key %s " % n["key"]
                    print "op %s " % n["op"]
                    print "value: %s " % n["value"]
                    print "-------------------------------"
                        
                return True
 
    def print_list(self,list):
        for item in list:
            if type(item) == type([]):
                self.print_list(item)
            elif type(item) == type({}):
                self.print_map(item)        
            else:
                print "%s" % item

    def print_map(self,map):
        for n,v in map.items():
            if type(v) == type([]):
                print "------%s start------" % n
                self.print_list(v)
                print "------%s ended------" % n
            elif type(v) == type({}):
                print "------%s start------" % n
                self.print_map(v)
                print "------%s ended------" % n
            else:
                print "%s = %-20s" % (n,v)

    def print_vmcfg_list(self,vmcfg_list):
        print "==========================="
        if type(vmcfg_list) == type([]):
            self.print_list(vmcfg_list)
        elif type(vmcfg_list) == type({}):
            self.print_map(vmcfg_list)    
        else:
            print "%s" % vmcfg_list

    def query_count(self,type,vt_name):
        '''显示指定的虚拟机配置信息  '''
        result = self.apisvr.tecs.vmtemplate.query(self.session,VT_START_INDEX,VT_QUERY_COUNT,type,vt_name) 
        if result[0] != 0:
            print "failed to query vmt information, type %d! error code =%d" % (type,result[0])
            print "error string = %s" % result[1]
            return None

        print "vmtemplate information:" 
        print "next_index = %d" % result[1]
        print "max_count = %d" % result[2]
        vmcfg_list = result[3]
        self.print_vmcfg_list(vmcfg_list)

        return result[2]

    def query_disks_count(self,type,vt_name):
        '''显示指定的虚拟机配置信息  '''
        result = self.apisvr.tecs.vmtemplate.query(self.session,VT_START_INDEX,VT_QUERY_COUNT,type,vt_name) 
        if result[0] != 0:
            print "failed to query vmt information, type %d! error code =%d" % (type,result[0])
            print "error string = %s" % result[1]
            return None
        else:
            count = 0;
            if len(result[3]) == 0:
                print "no vmt type %d!" % type
                return count;

            vmcfg_list = result[3]
            self.print_vmcfg_list(vmcfg_list)
            count = 0;
            for v in vmcfg_list:             
                disk_list = v["base_info"]["disks"]
                for d in disk_list:
                    count = count + 1
            print "count = %d" % count
            return count  
                    
    def query_nic_count(self,type,vt_name):
        '''显示指定的虚拟机配置信息  '''
        result = self.apisvr.tecs.vmtemplate.query(self.session,VT_START_INDEX,VT_QUERY_COUNT,type,vt_name) 
        if result[0] != 0:
            print "failed to query vmt information, type %d! error code =%d" % (type,result[0])
            print "error string = %s" % result[1]
            return None
        else:
            count = 0;
            if len(result[3]) == 0:
                print "no vmt type %d!" % type
                return count;
            
            vmcfg_list = result[3]
            self.print_vmcfg_list(vmcfg_list)

            for v in vmcfg_list:
                nic_list = v["base_info"]["nics"]
                for n in nic_list:
                    count = count + 1
            return count

    def query_value(self,type,vt_name,query_name):
        result = self.apisvr.tecs.vmtemplate.query(self.session,VT_START_INDEX,VT_QUERY_COUNT,type,vt_name) 
        if result[0] != 0:
            print "failed to query vmt information, type %d! error code =%d" % (type,result[0])
            print "error string = %s" % result[1]
            return None
        else:
            count = 0;
            if len(result[3]) == 0:
                print "no vmt type %d!" % type
                return None;
            
            vmcfg_list = result[3]
            self.print_vmcfg_list(vmcfg_list)

            for v in vmcfg_list:
                return v[query_name]
            return None


    def query_base_info_value(self,query_type,vt_name,query_name,query_name2="",which=0):
        result = self.apisvr.tecs.vmtemplate.query(self.session,VT_START_INDEX,VT_QUERY_COUNT,query_type,vt_name) 
        if result[0] != 0:
            print "failed to query vmt information, type %d! error code =%d" % (query_type,result[0])
            print "error string = %s" % result[1]
            return None
        else:
            count = 0;
            if len(result[3]) == 0:
                print "no vmt type %d!" % query_type
                return None;
            
            vmcfg_list = result[3]
            self.print_vmcfg_list(vmcfg_list)
            for v in vmcfg_list:
                v_type = type(v["base_info"][query_name])
                if v_type == type([]):
                    v_map_list = v["base_info"][query_name]
                    for v_map in v_map_list:
                        print which
                        print"v_map=%s" %v_map
                        if which == 0:
                            return v_map[query_name2]
                        else:
                            which = which -1
                elif v_type == type({}): 
                    v_map = v["base_info"][query_name]                    
                    return v_map[query_name2]
                else:                    
                    return v["base_info"][query_name]

            return None

    def query_pub(self,type,vt_name):
        return self.query_value(type,vt_name,"is_public")          
                   
    def query_vcpu(self,type,vt_name):
        return self.query_base_info_value(type,vt_name,"vcpu")
        
    def query_description(self,type,vt_name):
        return self.query_base_info_value(type,vt_name,"description")
        
    def query_vnc_passwd(self,type,vt_name):
        return self.query_base_info_value(type,vt_name,"vnc_passwd")
        
    def query_virt_type(self,type,vt_name):
        return self.query_base_info_value(type,vt_name,"virt_type")

    def query_memory(self,type,vt_name):
        return self.query_base_info_value(type,vt_name,"memory")     
    
    def query_vm_wdtime(self,type,vt_name):
        return self.query_base_info_value(type,vt_name,"vm_wdtime")   

    def query_machine_value(self,type,vt_name,query_name):
        return self.query_base_info_value(type,vt_name,"machine",query_name) 

    def query_machine_bus(self,type,vt_name):
        return self.query_machine_value(type,vt_name,"bus") 

    def query_disks_value(self,type,vt_name,query_name):
        return self.query_base_info_value(type,vt_name,"disks",query_name) 

    def query_disks_size(self,type,vt_name):
        return self.query_disks_value(type,vt_name,"size") 

    def query_disks_bus(self,type,vt_name):
        return self.query_disks_value(type,vt_name,"bus") 

    def query_disks_target(self,type,vt_name):
        return self.query_disks_value(type,vt_name,"target") 

    def query_disks_fstype(self,type,vt_name):
        return self.query_disks_value(type,vt_name,"fstype") 

    def query_nics_value(self,type,vt_name,query_name,which):
        return self.query_base_info_value(type,vt_name,"nics",query_name,which) 

    def query_nics_index(self,type,vt_name,which):
        return self.query_nics_value(type,vt_name,"nic_index",which) 
                    
    
